Error in inference for mlx-community/Nemotron-Cascade-2-30B-A3B-8bit

2026-03-28 22:52:34,612 [ERROR] olmlx.utils.streaming: Inference error (TypeError): create_attention_mask() missing 2 required positional arguments: 'return_array' and 'window_size'
Traceback (most recent call last):
  File "/Users/daniel/devel/olmlx-1/olmlx/utils/streaming.py", line 68, in _run
    for resp in gen:
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/generate.py", line 705, in stream_generate
    for n, (token, logprobs, from_draft) in enumerate(token_generator):
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/generate.py", line 694, in <genexpr>
    token_generator = (
                      ^
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/generate.py", line 429, in generate_step
    _model_call(
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/generate.py", line 390, in _model_call
    return model(input_tokens, cache=prompt_cache)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/models/nemotron_h.py", line 487, in __call__
    out = self.backbone(inputs, cache=cache)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/models/nemotron_h.py", line 455, in __call__
    ssm_mask = create_ssm_mask(hidden_states, cache[self.ssm_idx])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/daniel/devel/olmlx-1/.venv/lib/python3.11/site-packages/mlx_lm/models/base.py", line 60, in create_ssm_mask
    return cache.make_mask(h.shape[1])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/daniel/devel/olmlx-1/olmlx/engine/turboquant_cache.py", line 140, in make_mask
    return create_attention_mask(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: create_attention_mask() missing 2 required positional arguments: 'return_array' and 'window_size'

2026-03-28 22:52:34,614 [ERROR] olmlx.routers.anthropic: Error during Anthropic streaming: TypeError: create_attention_mask() missing 2 required positional arguments: 'return_array' and 'window_size'
Traceback (most recent call last):
  File "/Users/daniel/devel/olmlx-1/olmlx/routers/anthropic.py", line 694, in stream_sse
    async for event in path:
  File "/Users/daniel/devel/olmlx-1/olmlx/routers/anthropic.py", line 310, in _stream_buffered_with_tools
    async for chunk in _with_keepalive_pings(result, interval=KEEPALIVE_PING_INTERVAL):
  File "/Users/daniel/devel/olmlx-1/olmlx/routers/anthropic.py", line 248, in _with_keepalive_pings
    item = next_item_task.result()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/daniel/devel/olmlx-1/olmlx/engine/inference.py", line 1177, in _stream_completion
    async for token in stream:
  File "/Users/daniel/devel/olmlx-1/olmlx/utils/streaming.py", line 235, in __anext__
    raise RuntimeError(f"{exc_type}: {item[_ERROR_KEY]}")
RuntimeError: TypeError: create_attention_mask() missing 2 required positional arguments: 'return_array' and 'window_size'

--- a/olmlx/engine/turboquant_cache.py
+++ b/olmlx/engine/turboquant_cache.py
@@ -11,7 +11,7 @@ from typing import Any
 
 import mlx.core as mx
 
-from mlx_lm.models.cache import _BaseCache, create_attention_mask
+from mlx_lm.models.cache import KVCache, _BaseCache, create_attention_mask
 
 from olmlx.engine.turboquant import (
     TurboQuantRotation,
@@ -177,8 +177,13 @@ def _detect_head_dim(model: Any) -> int:
         ) from e
 
 
-def make_turboquant_cache(model: Any, bits: int) -> list[TurboQuantKVCache]:
-    """Create a list of TurboQuantKVCache objects, one per model layer."""
+def make_turboquant_cache(model: Any, bits: int) -> list:
+    """Create a cache list with TurboQuantKVCache for attention layers.
+
+    For hybrid models (e.g. Nemotron-H with SSM + attention layers), only
+    attention-layer caches (KVCache) are replaced with TurboQuantKVCache.
+    Non-attention caches (e.g. ArraysCache for SSM/Mamba layers) are preserved.
+    """
     num_layers = len(model.layers)
     head_dim = _detect_head_dim(model)
 
@@ -189,16 +194,31 @@ def make_turboquant_cache(model: Any, bits: int) -> list[TurboQuantKVCache]:
             f"got head_dim={head_dim}"
         )
 
+    # Get default cache layout from model if available (hybrid models
+    # return different cache types per layer, e.g. ArraysCache for SSM)
+    try:
+        default_caches = model.make_cache()
+        if not isinstance(default_caches, list) or len(default_caches) != num_layers:
+            default_caches = [None] * num_layers
+    except AttributeError:
+        default_caches = [None] * num_layers
+
     caches = []
-    for i in range(num_layers):
-        rot_k = TurboQuantRotation(head_dim=head_dim, seed=i * 2)
-        rot_v = TurboQuantRotation(head_dim=head_dim, seed=i * 2 + 1)
-        caches.append(
-            TurboQuantKVCache(bits=bits, rotation_key=rot_k, rotation_value=rot_v)
-        )
+    tq_count = 0
+    for i, default in enumerate(default_caches):
+        if default is None or isinstance(default, KVCache):
+            rot_k = TurboQuantRotation(head_dim=head_dim, seed=i * 2)
+            rot_v = TurboQuantRotation(head_dim=head_dim, seed=i * 2 + 1)
+            caches.append(
+                TurboQuantKVCache(bits=bits, rotation_key=rot_k, rotation_value=rot_v)
+            )
+            tq_count += 1
+        else:
+            caches.append(default)
 
     logger.info(
-        "Created TurboQuant KV cache: %d layers, %d-bit, head_dim=%d",
+        "Created TurboQuant KV cache: %d/%d layers quantized, %d-bit, head_dim=%d",
+        tq_count,
         num_layers,
         bits,
         head_dim,

--- a/olmlx/engine/turboquant_cache.py
+++ b/olmlx/engine/turboquant_cache.py
@@ -196,11 +196,11 @@ def make_turboquant_cache(model: Any, bits: int) -> list:
 
     # Get default cache layout from model if available (hybrid models
     # return different cache types per layer, e.g. ArraysCache for SSM)
-    try:
+    if hasattr(model, "make_cache"):
         default_caches = model.make_cache()
         if not isinstance(default_caches, list) or len(default_caches) != num_layers:
             default_caches = [None] * num_layers
-    except AttributeError:
+    else:
         default_caches = [None] * num_layers
 
     caches = []

--- a/tests/test_turboquant.py
+++ b/tests/test_turboquant.py
@@ -461,6 +461,53 @@ class TestMakeTurboQuantCache:
         cache = make_turboquant_cache(model, bits=4)
         assert cache[0].rotation_key.matrix.shape == (128, 128)
 
+    def test_hybrid_model_preserves_non_kv_caches(self):
+        """Hybrid models (e.g. Nemotron-H) have SSM layers needing ArraysCache."""
+        from unittest.mock import MagicMock
+
+        from mlx_lm.models.cache import ArraysCache, KVCache
+
+        from olmlx.engine.turboquant_cache import (
+            TurboQuantKVCache,
+            make_turboquant_cache,
+        )
+
+        model = MagicMock()
+        model.layers = [MagicMock() for _ in range(4)]
+        model.args.head_dim = 64
+        # Simulate Nemotron-H: SSM, attention, SSM, attention
+        model.make_cache.return_value = [
+            ArraysCache(size=2),
+            KVCache(),
+            ArraysCache(size=2),
+            KVCache(),
+        ]
+
+        cache = make_turboquant_cache(model, bits=4)
+        assert len(cache) == 4
+        assert isinstance(cache[0], ArraysCache)
+        assert isinstance(cache[1], TurboQuantKVCache)
+        assert isinstance(cache[2], ArraysCache)
+        assert isinstance(cache[3], TurboQuantKVCache)
+
+    def test_model_without_make_cache_all_turboquant(self):
+        """Models without make_cache() get TurboQuantKVCache for all layers."""
+        from unittest.mock import MagicMock
+
+        from olmlx.engine.turboquant_cache import (
+            TurboQuantKVCache,
+            make_turboquant_cache,
+        )
+
+        model = MagicMock(spec=[])  # no make_cache attribute
+        model.layers = [MagicMock() for _ in range(3)]
+        model.args = MagicMock()
+        model.args.head_dim = 64
+
+        cache = make_turboquant_cache(model, bits=4)
+        assert len(cache) == 3
+        assert all(isinstance(c, TurboQuantKVCache) for c in cache)
+
 
 # ---------------------------------------------------------------------------
 # Config tests

